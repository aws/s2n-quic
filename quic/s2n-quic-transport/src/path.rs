//! This module contains the Manager implementation

use crate::{connection::PeerIdRegistry, space::EARLY_ACK_SETTINGS, transmission};
use s2n_quic_core::{
    connection,
    inet::{DatagramInfo, SocketAddress},
    recovery::{CongestionController, RTTEstimator},
    transport::error::TransportError,
};
use smallvec::SmallVec;

use crate::transmission::{Interest, WriteContext};
use s2n_quic_core::ack_set::AckSet;
/// re-export core
pub use s2n_quic_core::path::*;

/// The amount of Paths that can be maintained without using the heap
const INLINE_PATH_LEN: usize = 5;

/// The PathManager handles paths for a specific connection.
/// It will handle path validation operations, and track the active path for a connection.
#[derive(Debug)]
pub struct Manager<CC: CongestionController> {
    /// Path array
    paths: SmallVec<[Path<CC>; INLINE_PATH_LEN]>,

    /// Registry of `connection::PeerId`s
    peer_id_registry: PeerIdRegistry,

    /// Index to the active path
    active: usize,
}

impl<CC: CongestionController> Manager<CC> {
    pub fn new(initial_path: Path<CC>, peer_id_registry: PeerIdRegistry) -> Self {
        Manager {
            paths: SmallVec::from_elem(initial_path, 1),
            peer_id_registry,
            active: 0,
        }
    }

    /// Return the active path
    pub fn active_path(&self) -> &Path<CC> {
        &self.paths[self.active]
    }

    /// Return a mutable reference to the active path
    pub fn active_path_mut(&mut self) -> &mut Path<CC> {
        &mut self.paths[self.active]
    }

    /// Return the Id of the active path
    pub fn active_path_id(&self) -> Id {
        Id(self.active)
    }

    /// Returns the Path for the provided address if the PathManager knows about it
    pub fn path(&self, peer_address: &SocketAddress) -> Option<(Id, &Path<CC>)> {
        self.paths
            .iter()
            .enumerate()
            .find(|(_id, path)| *peer_address == path.peer_socket_address)
            .map(|(id, path)| (Id(id), path))
    }

    /// Returns the Path for the provided address if the PathManager knows about it
    pub fn path_mut(&mut self, peer_address: &SocketAddress) -> Option<(Id, &mut Path<CC>)> {
        self.paths
            .iter_mut()
            .enumerate()
            .find(|(_id, path)| *peer_address == path.peer_socket_address)
            .map(|(id, path)| (Id(id), path))
    }

    /// Called when a datagram is received on a connection
    /// Upon success, returns a `(Id, bool)` containing the path ID and a boolean that is
    /// true if the path had been amplification limited prior to receiving the datagram
    /// and is now no longer amplification limited.
    #[allow(unreachable_code)]
    #[allow(unused_variables)]
    pub fn on_datagram_received<NewCC: FnOnce() -> CC>(
        &mut self,
        datagram: &DatagramInfo,
        is_handshake_confirmed: bool,
        new_congestion_controller: NewCC,
    ) -> Result<(Id, bool), TransportError> {
        if let Some((id, path)) = self.path_mut(&datagram.remote_address) {
            let unblocked = path.on_bytes_received(datagram.payload_len);
            return Ok((id, unblocked));
        }

        //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#9.5
        //= type=TODO
        //= tracking-issue=316
        //# Similarly, an endpoint MUST NOT reuse a connection ID when sending to
        //# more than one destination address.

        //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#9.5
        //= type=TODO
        //= tracking-issue=316
        //# Due to network changes outside
        //# the control of its peer, an endpoint might receive packets from a new
        //# source address with the same destination connection ID, in which case
        //# it MAY continue to use the current connection ID with the new remote
        //# address while still sending from the same local address.

        //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#9
        //# The design of QUIC relies on endpoints retaining a stable address
        //# for the duration of the handshake.  An endpoint MUST NOT initiate
        //# connection migration before the handshake is confirmed, as defined
        //# in section 4.1.2 of [QUIC-TLS].
        if !is_handshake_confirmed {
            return Err(TransportError::PROTOCOL_VIOLATION);
        }

        // Since we are not currently supporting connection migration (whether it was deliberate or
        // not), we will error our at this point to avoid re-using a peer connection ID.
        // TODO: This would be better handled as a stateless reset so the peer can terminate the
        //       connection immediately. https://github.com/awslabs/s2n-quic/issues/317
        return Err(TransportError::INTERNAL_ERROR);

        let path = Path::new(
            datagram.remote_address,
            // TODO https://github.com/awslabs/s2n-quic/issues/316
            // The existing peer connection id may only be reused if the destination
            // connection ID on this packet had not been used before (this would happen
            // when the peer's remote address gets changed due to circumstances out of their
            // control). Otherwise we will need to consume a new connection::PeerId by calling
            // PeerIdRegistry::consume_new_id_if_necessary(None) and ignoring the request if
            // no new connection::PeerId is available to use.
            self.active_path().peer_connection_id,
            RTTEstimator::new(EARLY_ACK_SETTINGS.max_ack_delay),
            new_congestion_controller(),
            true,
        );
        let id = Id(self.paths.len());
        self.paths.push(path);
        Ok((id, false))
    }

    /// Writes any frames the path manager wishes to transmit to the given context
    pub fn on_transmit<W: WriteContext>(&mut self, context: &mut W) {
        self.peer_id_registry.on_transmit(context)
    }

    /// Called when packets are acknowledged
    pub fn on_packet_ack<A: AckSet>(&mut self, ack_set: &A) {
        self.peer_id_registry.on_packet_ack(ack_set)
    }

    /// Called when packets are lost
    pub fn on_packet_loss<A: AckSet>(&mut self, ack_set: &A) {
        self.peer_id_registry.on_packet_loss(ack_set)
    }

    //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#8.2.2
    //= type=TODO
    //= tracking-issue=404
    //= feature=Client connection migration
    //# On receiving a PATH_CHALLENGE frame, an endpoint MUST respond by
    //# echoing the data contained in the PATH_CHALLENGE frame in a
    //# PATH_RESPONSE frame.
    pub fn on_path_challenge(
        &mut self,
        _peer_address: &SocketAddress,
        _challenge: s2n_quic_core::frame::PathChallenge,
    ) {
        //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#8.2.2
        //= type=TODO
        //= tracking-issue=406
        //= feature=Connection migration
        //# An endpoint MUST NOT delay transmission of a
        //# packet containing a PATH_RESPONSE frame unless constrained by
        //# congestion control.

        //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#8.2.2
        //= type=TODO
        //= tracking-issue=407
        //= feature=Connection migration
        //# An endpoint MUST NOT send more than one PATH_RESPONSE frame in
        //# response to one PATH_CHALLENGE frame; see Section 13.3.
    }

    //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#8.2.3
    //# Path validation succeeds when a PATH_RESPONSE frame is received that
    //# contains the data that was sent in a previous PATH_CHALLENGE frame.
    //# A PATH_RESPONSE frame received on any network path validates the path
    //# on which the PATH_CHALLENGE was sent.
    pub fn on_path_response(
        &mut self,
        peer_address: &SocketAddress,
        response: s2n_quic_core::frame::PathResponse,
    ) {
        if let Some((_id, path)) = self.path_mut(peer_address) {
            // We may have received a duplicate packet, only call the on_validated handler
            // one time.
            if path.is_validated() {
                return;
            }

            if let Some(expected_response) = path.challenge {
                if &expected_response == response.data {
                    path.on_validated();
                }
            }
        }
    }

    /// Called when a token is received that was issued in a Retry packet
    pub fn on_retry_token(&self, _peer_address: &SocketAddress, _token: &[u8]) {}

    /// Called when a token is received that was issued in a NEW_TOKEN frame
    pub fn on_new_token(&self, _peer_address: &SocketAddress, _token: &[u8]) {}

    /// Start the validation process for a path
    pub fn validate_path(&self, _path: Path<CC>) {}

    //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#10.3
    //# Tokens are
    //# invalidated when their associated connection ID is retired via a
    //# RETIRE_CONNECTION_ID frame (Section 19.16).
    pub fn on_connection_id_retire(&self, _connection_id: &connection::LocalId) {
        // TODO invalidate any tokens issued under this connection id
    }

    /// Called when a NEW_CONNECTION_ID frame is received from the peer
    pub fn on_new_connection_id(
        &mut self,
        connection_id: &connection::PeerId,
        sequence_number: u32,
        retire_prior_to: u32,
        stateless_reset_token: &[u8],
    ) -> Result<(), TransportError> {
        // Register the new connection ID
        self.peer_id_registry.on_new_connection_id(
            connection_id,
            sequence_number,
            retire_prior_to,
            stateless_reset_token,
        )?;

        // TODO This new connection ID may retire IDs in use by multiple paths. Since we are not
        //      currently supporting connection migration, there is only one path, but once there
        //      are more than one we should decide what to do if there aren't enough new connection
        //      IDs available for all paths.
        //      See https://github.com/awslabs/s2n-quic/issues/358
        //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#5.1.2
        //# Upon receipt of an increased Retire Prior To field, the peer MUST
        //# stop using the corresponding connection IDs and retire them with
        //# RETIRE_CONNECTION_ID frames before adding the newly provided
        //# connection ID to the set of active connection IDs.
        // Ensure all paths are not using a newly retired connection ID
        for path in self.paths.iter_mut() {
            path.peer_connection_id = self
                .peer_id_registry
                .consume_new_id_if_necessary(Some(&path.peer_connection_id))
                .expect(
                    "There is only one path maintained currently and since a new ID was \
                delivered, there will always be a new ID available to consume if necessary",
                );
        }

        Ok(())
    }

    pub fn on_packet_received(&mut self) {}

    pub fn on_timeout(&mut self) {
        //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#8.2.4
        //= type=TODO
        //= tracking-issue=412
        //= feature=Connection migration
        //# Endpoints SHOULD abandon path validation based on a timer.

        //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#8.2.4
        //= type=TODO
        //= tracking-issue=412
        //= feature=Connection migration
        //# A value of
        //# three times the larger of the current Probe Timeout (PTO) or the
        //# initial timeout (that is, 2*kInitialRtt) as defined in
        //# [QUIC-RECOVERY] is RECOMMENDED.
    }
}

impl<CC: CongestionController> transmission::interest::Provider for Manager<CC> {
    // TODO Add transmission interests for path probes
    fn transmission_interest(&self) -> Interest {
        self.peer_id_registry.transmission_interest()
    }
}

/// Internal Id of a path in the manager
#[derive(Clone, Copy, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Id(usize);

impl<CC: CongestionController> core::ops::Index<Id> for Manager<CC> {
    type Output = Path<CC>;

    fn index(&self, id: Id) -> &Self::Output {
        &self.paths[id.0]
    }
}

impl<CC: CongestionController> core::ops::IndexMut<Id> for Manager<CC> {
    fn index_mut(&mut self, id: Id) -> &mut Self::Output {
        &mut self.paths[id.0]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use core::time::Duration;
    use s2n_quic_core::{
        frame::new_connection_id::STATELESS_RESET_TOKEN_LEN,
        inet::{DatagramInfo, ExplicitCongestionNotification},
        recovery::{congestion_controller::testing::Unlimited, RTTEstimator},
        time::Timestamp,
    };
    use std::net::SocketAddr;

    #[test]
    fn get_path_by_address_test() {
        let conn_id = connection::PeerId::try_from_bytes(&[0, 1, 2, 3, 4, 5]).unwrap();
        let first_path = Path::new(
            SocketAddress::default(),
            conn_id,
            RTTEstimator::new(Duration::from_millis(30)),
            Unlimited::default(),
            false,
        );

        let peer_id_registry = PeerIdRegistry::new(first_path.peer_connection_id, None);
        let manager = Manager::new(first_path, peer_id_registry);
        assert_eq!(manager.paths.len(), 1);

        let (_id, matched_path) = manager.path(&SocketAddress::default()).unwrap();
        assert_eq!(matched_path, &first_path);
    }

    #[test]
    fn path_validate_test() {
        let first_conn_id = connection::PeerId::try_from_bytes(&[0, 1, 2, 3, 4, 5]).unwrap();
        let mut first_path = Path::new(
            SocketAddress::default(),
            first_conn_id,
            RTTEstimator::new(Duration::from_millis(30)),
            Unlimited::default(),
            false,
        );
        first_path.challenge = Some([0u8; 8]);

        let peer_id_registry = PeerIdRegistry::new(first_path.peer_connection_id, None);
        let mut manager = Manager::new(first_path, peer_id_registry);
        assert_eq!(manager.paths.len(), 1);
        {
            let (_id, first_path) = manager.path(&first_path.peer_socket_address).unwrap();
            assert_eq!(first_path.is_validated(), false);
        }

        let frame = s2n_quic_core::frame::PathResponse { data: &[0u8; 8] };
        manager.on_path_response(&first_path.peer_socket_address, frame);
        {
            let (_id, first_path) = manager.path(&first_path.peer_socket_address).unwrap();
            assert_eq!(first_path.is_validated(), true);
        }
    }

    #[test]
    #[allow(unreachable_code)]
    fn new_peer_test() {
        let first_conn_id = connection::PeerId::try_from_bytes(&[0, 1, 2, 3, 4, 5]).unwrap();
        let first_path = Path::new(
            SocketAddress::default(),
            first_conn_id,
            RTTEstimator::new(Duration::from_millis(30)),
            Unlimited::default(),
            false,
        );
        let peer_id_registry = PeerIdRegistry::new(first_path.peer_connection_id, None);
        let mut manager = Manager::new(first_path, peer_id_registry);
        assert_eq!(manager.paths.len(), 1);
        let new_addr: SocketAddr = "127.0.0.1:80".parse().unwrap();
        let new_addr = SocketAddress::from(new_addr);

        assert_eq!(manager.path(&SocketAddress::default()).is_some(), true);
        assert_eq!(manager.path(&new_addr), None);
        assert_eq!(manager.paths.len(), 1);

        let datagram = DatagramInfo {
            timestamp: unsafe { Timestamp::from_duration(Duration::from_millis(30)) },
            remote_address: new_addr,
            payload_len: 0,
            ecn: ExplicitCongestionNotification::default(),
            destination_connection_id: connection::LocalId::TEST_ID,
        };

        assert_eq!(
            Err(TransportError::INTERNAL_ERROR),
            manager.on_datagram_received(&datagram, true, Unlimited::default)
        );

        return;

        // TODO: The following code can be enabled once connection migration is implemented
        assert_eq!(manager.path(&new_addr).is_some(), true);
        assert_eq!(manager.paths.len(), 2);

        let new_addr: SocketAddr = "127.0.0.1:443".parse().unwrap();
        let new_addr = SocketAddress::from(new_addr);
        let datagram = DatagramInfo {
            timestamp: unsafe { Timestamp::from_duration(Duration::from_millis(30)) },
            remote_address: new_addr,
            payload_len: 0,
            ecn: ExplicitCongestionNotification::default(),
            destination_connection_id: connection::LocalId::TEST_ID,
        };

        assert_eq!(
            manager
                .on_datagram_received(&datagram, false, Unlimited::default)
                .is_err(),
            true
        );
        assert_eq!(manager.paths.len(), 2);
    }

    //= https://tools.ietf.org/id/draft-ietf-quic-transport-32.txt#5.1.2
    //= type=test
    //# Upon receipt of an increased Retire Prior To field, the peer MUST
    //# stop using the corresponding connection IDs and retire them with
    //# RETIRE_CONNECTION_ID frames before adding the newly provided
    //# connection ID to the set of active connection IDs.
    #[test]
    fn stop_using_a_retired_connection_id() {
        let id_1 = connection::PeerId::try_from_bytes(b"id01").unwrap();
        let first_path = Path::new(
            SocketAddress::default(),
            id_1,
            RTTEstimator::new(Duration::from_millis(30)),
            Unlimited::default(),
            false,
        );
        let peer_id_registry = PeerIdRegistry::new(first_path.peer_connection_id, None);
        let mut manager = Manager::new(first_path, peer_id_registry);

        let id_2 = connection::PeerId::try_from_bytes(b"id02").unwrap();
        assert!(manager
            .on_new_connection_id(&id_2, 1, 1, &[1_u8; STATELESS_RESET_TOKEN_LEN])
            .is_ok());

        assert_eq!(id_2, manager.paths[0].peer_connection_id);
    }
}
