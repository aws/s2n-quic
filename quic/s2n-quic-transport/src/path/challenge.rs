// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{contexts::WriteContext, transmission};
use s2n_quic_core::{
    ct::ConstantTimeEq,
    frame,
    time::{timer, Duration, Timer, Timestamp},
};

pub type Data = [u8; frame::path_challenge::DATA_LEN];
const DISABLED_DATA: Data = [0; frame::path_challenge::DATA_LEN];

#[derive(Clone, Debug)]
pub struct Challenge {
    state: State,
    abandon_duration: Duration,
    abandon_timer: Timer,
    data: Data,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum State {
    /// PATH_CHALLENGE is not used for path validation when initiating a new connection
    InitialPathDisabled,

    /// A Challenge frame must be sent. The `u8` represents the remaining number of retries
    RequiresTransmission(u8),

    /// Challenge has been sent and we are awaiting a response until the abandon timer expires
    PendingResponse,

    /// The Challenge has been abandoned due to the abandon_timer
    Abandoned,

    /// When the PATH_CHALLENGE was validated by a PATH_RESPONSE
    Validated,
}

impl transmission::interest::Provider for State {
    #[inline]
    fn transmission_interest<Q: transmission::interest::Query>(
        &self,
        query: &mut Q,
    ) -> transmission::interest::Result {
        match self {
            State::RequiresTransmission(_) => query.on_interest(transmission::Interest::NewData),
            _ => Ok(()),
        }
    }
}

impl Challenge {
    pub fn new(abandon_duration: Duration, data: Data) -> Self {
        Self {
            //= https://www.rfc-editor.org/rfc/rfc9000.txt#8.2.1
            //# An endpoint SHOULD NOT probe a new path with packets containing a
            //# PATH_CHALLENGE frame more frequently than it would send an Initial
            //# packet.

            //= https://www.rfc-editor.org/rfc/rfc9000.txt#8.2.1
            //# An endpoint MAY send multiple PATH_CHALLENGE frames to guard against
            //# packet loss.

            // Re-transmitting twice guards against packet loss, while remaining
            // below the amplification limit of 3.
            state: State::RequiresTransmission(2),
            abandon_duration,
            abandon_timer: Timer::default(),
            data,
        }
    }

    pub fn disabled() -> Self {
        Self {
            state: State::InitialPathDisabled,
            abandon_duration: Duration::ZERO,
            abandon_timer: Timer::default(),
            data: DISABLED_DATA,
        }
    }

    /// When a PATH_CHALLENGE is transmitted this handles any internal state operations.
    pub fn on_transmit<W: WriteContext>(&mut self, context: &mut W) {
        match self.state {
            State::RequiresTransmission(0) => self.state = State::PendingResponse,
            State::RequiresTransmission(remaining) => {
                //= https://www.rfc-editor.org/rfc/rfc9000.txt#8.2.1
                //# However, an endpoint SHOULD NOT send multiple
                //# PATH_CHALLENGE frames in a single packet.
                let frame = frame::PathChallenge { data: &self.data };

                if context.write_frame(&frame).is_some() {
                    let remaining = remaining - 1;
                    self.state = State::RequiresTransmission(remaining);

                    if !self.abandon_timer.is_armed() {
                        self.abandon_timer
                            .set(context.current_time() + self.abandon_duration);
                    }
                }
            }
            _ => {}
        }
    }

    pub fn on_timeout(&mut self, timestamp: Timestamp) {
        if self.abandon_timer.poll_expiration(timestamp).is_ready() {
            self.abandon();
        }
    }

    pub fn abandon(&mut self) {
        if self.is_pending() {
            self.state = State::Abandoned;
            self.abandon_timer.cancel();
        }
    }

    pub fn is_disabled(&self) -> bool {
        matches!(self.state, State::InitialPathDisabled)
    }

    pub fn is_pending(&self) -> bool {
        matches!(
            self.state,
            State::PendingResponse | State::RequiresTransmission(_)
        )
    }

    pub fn on_validated(&mut self, data: &[u8]) -> bool {
        if self.is_pending() && ConstantTimeEq::ct_eq(&self.data[..], data).into() {
            self.state = State::Validated;
            true
        } else {
            false
        }
    }
}

impl timer::Provider for Challenge {
    #[inline]
    fn timers<Q: timer::Query>(&self, query: &mut Q) -> timer::Result {
        self.abandon_timer.timers(query)?;

        Ok(())
    }
}

impl transmission::interest::Provider for Challenge {
    #[inline]
    fn transmission_interest<Q: transmission::interest::Query>(
        &self,
        query: &mut Q,
    ) -> transmission::interest::Result {
        self.state.transmission_interest(query)
    }
}

#[cfg(any(test, feature = "testing"))]
pub mod testing {
    use super::*;
    use s2n_quic_core::time::{Clock, Duration, NoopClock};

    pub fn helper_challenge() -> Helper {
        let now = NoopClock {}.get_time();
        let abandon_duration = Duration::from_millis(10_000);
        let expected_data: [u8; 8] = [0; 8];

        let challenge = Challenge::new(abandon_duration, expected_data);

        Helper {
            now,
            abandon_duration,
            expected_data,
            challenge,
        }
    }

    #[allow(dead_code)]
    pub struct Helper {
        pub now: Timestamp,
        pub abandon_duration: Duration,
        pub expected_data: Data,
        pub challenge: Challenge,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::contexts::testing::{MockWriteContext, OutgoingFrameBuffer};
    use s2n_quic_core::{
        endpoint,
        time::{Clock, Duration, NoopClock},
    };
    use testing::*;

    //= https://www.rfc-editor.org/rfc/rfc9000.txt#8.2.1
    //= type=test
    //# An endpoint MAY send multiple PATH_CHALLENGE frames to guard against
    //# packet loss.
    #[test]
    fn create_challenge_that_requires_two_transmissions() {
        let helper = helper_challenge();
        assert_eq!(helper.challenge.state, State::RequiresTransmission(2));
    }

    #[test]
    fn create_disabled_challenge() {
        let challenge = Challenge::disabled();
        assert_eq!(challenge.state, State::InitialPathDisabled);
        assert!(!challenge.abandon_timer.is_armed());
    }

    //= https://www.rfc-editor.org/rfc/rfc9000.txt#8.2.1
    //= type=test
    //# An endpoint SHOULD NOT probe a new path with packets containing a
    //# PATH_CHALLENGE frame more frequently than it would send an Initial
    //# packet.
    #[test]
    fn transmit_challenge_only_twice() {
        // Setup:
        let mut helper = helper_challenge();
        let mut frame_buffer = OutgoingFrameBuffer::new();
        let mut context = MockWriteContext::new(
            helper.now,
            &mut frame_buffer,
            transmission::Constraint::None,
            transmission::Mode::Normal,
            endpoint::Type::Client,
        );
        assert_eq!(helper.challenge.state, State::RequiresTransmission(2));

        // Trigger:
        helper.challenge.on_transmit(&mut context);

        // Expectation:
        //= https://www.rfc-editor.org/rfc/rfc9000.txt#8.2.1
        //= type=test
        //# However, an endpoint SHOULD NOT send multiple
        //# PATH_CHALLENGE frames in a single packet.
        assert_eq!(context.frame_buffer.len(), 1);

        assert_eq!(helper.challenge.state, State::RequiresTransmission(1));
        let written_data = match context.frame_buffer.pop_front().unwrap().as_frame() {
            frame::Frame::PathChallenge(frame) => Some(*frame.data),
            _ => None,
        }
        .unwrap();
        assert_eq!(written_data, helper.expected_data);

        // Trigger:
        helper.challenge.on_transmit(&mut context);

        // Expectation:
        assert_eq!(helper.challenge.state, State::RequiresTransmission(0));
        let written_data = match context.frame_buffer.pop_front().unwrap().as_frame() {
            frame::Frame::PathChallenge(frame) => Some(*frame.data),
            _ => None,
        }
        .unwrap();
        assert_eq!(written_data, helper.expected_data);

        // Trigger:
        helper.challenge.on_transmit(&mut context);

        // Expectation:
        assert_eq!(helper.challenge.state, State::PendingResponse);
        assert_eq!(context.frame_buffer.len(), 0);
    }

    #[test]
    fn successful_on_transmit_arms_the_timer() {
        // Setup:
        let mut helper = helper_challenge();
        let mut frame_buffer = OutgoingFrameBuffer::new();
        let mut context = MockWriteContext::new(
            helper.now,
            &mut frame_buffer,
            transmission::Constraint::None,
            transmission::Mode::Normal,
            endpoint::Type::Client,
        );
        assert_eq!(helper.challenge.state, State::RequiresTransmission(2));
        assert!(!helper.challenge.abandon_timer.is_armed());

        // Trigger:
        helper.challenge.on_transmit(&mut context);

        // Expectation:
        assert!(helper.challenge.abandon_timer.is_armed());
    }

    #[test]
    fn maintain_idle_and_dont_transmit_when_pending_response_state() {
        // Setup:
        let mut helper = helper_challenge();
        let mut frame_buffer = OutgoingFrameBuffer::new();
        let mut context = MockWriteContext::new(
            helper.now,
            &mut frame_buffer,
            transmission::Constraint::None,
            transmission::Mode::Normal,
            endpoint::Type::Client,
        );
        helper.challenge.state = State::PendingResponse;
        assert_eq!(helper.challenge.state, State::PendingResponse);

        // Trigger:
        helper.challenge.on_transmit(&mut context);

        // Expectation:
        assert_eq!(helper.challenge.state, State::PendingResponse);
        assert_eq!(context.frame_buffer.len(), 0);
    }

    #[test]
    fn test_on_timeout() {
        let mut helper = helper_challenge();
        let expiration_time = helper.now + helper.abandon_duration;

        let mut frame_buffer = OutgoingFrameBuffer::new();
        let mut context = MockWriteContext::new(
            helper.now,
            &mut frame_buffer,
            transmission::Constraint::None,
            transmission::Mode::Normal,
            endpoint::Type::Client,
        );
        helper.challenge.on_transmit(&mut context);

        helper
            .challenge
            .on_timeout(expiration_time - Duration::from_millis(10));
        assert!(helper.challenge.is_pending());

        helper
            .challenge
            .on_timeout(expiration_time + Duration::from_millis(10));
        assert!(!helper.challenge.is_pending());
    }

    #[test]
    fn challenge_must_remains_abandoned_once_abandoned() {
        let mut helper = helper_challenge();
        let expiration_time = helper.now + helper.abandon_duration;

        let mut frame_buffer = OutgoingFrameBuffer::new();
        let mut context = MockWriteContext::new(
            helper.now,
            &mut frame_buffer,
            transmission::Constraint::None,
            transmission::Mode::Normal,
            endpoint::Type::Client,
        );
        helper.challenge.on_transmit(&mut context);

        // Trigger:
        helper
            .challenge
            .on_timeout(expiration_time + Duration::from_millis(10));

        // Expectation:
        assert!(!helper.challenge.is_pending());

        // Trigger:
        helper
            .challenge
            .on_timeout(expiration_time - Duration::from_millis(10));

        // Expectation:
        assert!(!helper.challenge.is_pending());
    }

    #[test]
    fn dont_abandon_disabled_state() {
        let mut challenge = Challenge::disabled();
        let now = NoopClock {}.get_time();

        let mut frame_buffer = OutgoingFrameBuffer::new();
        let mut context = MockWriteContext::new(
            now,
            &mut frame_buffer,
            transmission::Constraint::None,
            transmission::Mode::Normal,
            endpoint::Type::Client,
        );
        challenge.on_transmit(&mut context);

        assert_eq!(challenge.state, State::InitialPathDisabled);

        let large_expiration_time = now + Duration::from_secs(1_000_000);
        challenge.on_timeout(large_expiration_time);
        assert_eq!(challenge.state, State::InitialPathDisabled);
    }

    #[test]
    fn test_is_abandoned() {
        let mut helper = helper_challenge();
        let expiration_time = helper.now + helper.abandon_duration;

        let mut frame_buffer = OutgoingFrameBuffer::new();
        let mut context = MockWriteContext::new(
            helper.now,
            &mut frame_buffer,
            transmission::Constraint::None,
            transmission::Mode::Normal,
            endpoint::Type::Client,
        );
        helper.challenge.on_transmit(&mut context);

        assert!(helper.challenge.is_pending());

        helper
            .challenge
            .on_timeout(expiration_time + Duration::from_millis(10));
        assert!(!helper.challenge.is_pending());
    }

    #[test]
    fn test_on_validate() {
        let mut helper = helper_challenge();

        let wrong_data: [u8; 8] = [5; 8];
        assert!(!helper.challenge.on_validated(&wrong_data));
        assert!(helper.challenge.is_pending());

        assert!(helper.challenge.on_validated(&helper.expected_data));
        assert_eq!(helper.challenge.state, State::Validated);
    }

    #[test]
    fn is_disabled() {
        let challenge = Challenge::disabled();

        assert_eq!(challenge.state, State::InitialPathDisabled);
        assert!(challenge.is_disabled());
    }

    #[test]
    fn dont_validate_disabled_state() {
        let mut helper = helper_challenge();
        helper.challenge.state = State::InitialPathDisabled;

        assert!(!helper.challenge.on_validated(&helper.expected_data));
        assert_eq!(helper.challenge.state, State::InitialPathDisabled);
    }

    #[test]
    fn dont_abandon_a_validated_challenge() {
        let mut helper = helper_challenge();
        helper.challenge.state = State::Validated;

        helper.challenge.abandon();

        assert_eq!(helper.challenge.state, State::Validated);
    }

    #[test]
    fn cancel_abandon_timer_on_abandon() {
        let mut helper = helper_challenge();

        helper.challenge.abandon();

        assert!(!helper.challenge.abandon_timer.is_armed());
    }
}
