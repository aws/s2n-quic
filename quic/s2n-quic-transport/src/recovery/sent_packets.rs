// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::path;
use core::convert::TryInto;
use s2n_quic_core::{
    frame::ack_elicitation::AckElicitation, inet::ExplicitCongestionNotification,
    packet::number::Map as PacketNumberMap, time::Timestamp,
};

//= https://tools.ietf.org/id/draft-ietf-quic-recovery-32.txt#A.1

//= https://tools.ietf.org/id/draft-ietf-quic-recovery-32.txt#A.1.1

pub type SentPackets = PacketNumberMap<SentPacketInfo>;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct SentPacketInfo {
    /// Indicates whether the packet counts towards bytes in flight
    pub congestion_controlled: bool,
    /// The number of bytes sent in the packet, not including UDP or IP overhead,
    /// but including QUIC framing overhead
    pub sent_bytes: u16,
    /// The time the packet was sent
    pub time_sent: Timestamp,
    /// Indicates whether a packet is ack-eliciting
    pub ack_elicitation: AckElicitation,
    /// The ID of the Path the packet was sent on
    pub path_id: path::Id,
    /// The ECN marker (if any) sent on the datagram that contained this packet
    pub ecn: ExplicitCongestionNotification,
}

impl SentPacketInfo {
    pub fn new(
        congestion_controlled: bool,
        sent_bytes: usize,
        time_sent: Timestamp,
        ack_elicitation: AckElicitation,
        path_id: path::Id,
        ecn: ExplicitCongestionNotification,
    ) -> Self {
        debug_assert_eq!(
            sent_bytes > 0,
            congestion_controlled,
            "sent bytes should be zero for packets that are not congestion controlled"
        );

        SentPacketInfo {
            congestion_controlled,
            sent_bytes: sent_bytes
                .try_into()
                .expect("sent_bytes exceeds max UDP payload size"),
            time_sent,
            ack_elicitation,
            path_id,
            ecn,
        }
    }
}

#[cfg(test)]
mod test {
    use crate::{path, recovery::SentPacketInfo};
    use s2n_quic_core::{
        frame::ack_elicitation::AckElicitation, inet::ExplicitCongestionNotification,
    };

    #[test]
    #[should_panic]
    fn too_large_packet() {
        SentPacketInfo::new(
            true,
            u16::max_value() as usize + 1,
            s2n_quic_platform::time::now(),
            AckElicitation::Eliciting,
            path::Id::new(0),
            ExplicitCongestionNotification::default(),
        );
    }

    #[test]
    fn sent_packet_info_size_test() {
        insta::assert_debug_snapshot!(
            stringify!(sent_packet_info_size_test),
            core::mem::size_of::<SentPacketInfo>()
        );

        assert_eq!(
            core::mem::size_of::<Option<SentPacketInfo>>(),
            core::mem::size_of::<SentPacketInfo>()
        );
    }
}
