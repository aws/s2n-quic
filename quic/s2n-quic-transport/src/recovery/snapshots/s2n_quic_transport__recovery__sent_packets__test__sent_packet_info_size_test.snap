---
source: quic/s2n-quic-transport/src/recovery/sent_packets.rs
expression: "core::mem::size_of::<SentPacketInfo>()"

---
16
